Add prototype for ginkgo solver integration

fritzgoebel opened the pull request:
This PR adds a prototype for the linear solver interface to the [Ginkgo](https://github.com/ginkgo-project/ginkgo) library 

To get this running, you need to:
- Install ginkgo from the [GLU integration branch](https://github.com/ginkgo-project/ginkgo/tree/glu) with CUDA enabled (see the [Installation pages](https://github.com/ginkgo-project/ginkgo/blob/glu/INSTALL.md)
- when building hiop, use the `HIOP_WITH_GINKGO` option and set `Ginkgo_DIR` to the path of your installation

Since the GLU integration only serves as a thankful starting point for the direct solver design in ginkgo, please regard this PR more as an interface suggestion than a stable solver integration.

pelesh commented:
@fritzgoebel, it might be a good idea to show in this PR how you use sparse examples 6 and 7, i.e. to add an option to these examples to use Ginkgo.

Please make sure following two options are enabled in sparse examples when using Ginkgo with HiOp:
```c++
    // Only linear solver runs on GPU
    nlp.options->SetStringValue("compute_mode", "hybrid");
    // LU solver needs to use inertia free approach
    nlp.options->SetStringValue("fact_acceptor", "inertia_free");
```

ashermancinelli commented on src/LinAlg/hiopLinSolverSparseGinkgo.cpp:
```diff
@@ -0,0 +1,376 @@
+
+  hiopLinSolverIndefSparseGinkgo::~hiopLinSolverIndefSparseGinkgo()
+  {
+    if(index_covert_CSR2Triplet_)
+      delete [] index_covert_CSR2Triplet_;
+    if(index_covert_extra_Diag2CSR_)
+      delete [] index_covert_extra_Diag2CSR_;
```
[Deleting a nullptr is a safe no-op, here's an example.](https://godbolt.org/z/Ks8oh55W9)

[From the standard:](https://en.cppreference.com/w/cpp/language/delete)

> If expression evaluates to a null pointer value, no destructors are called, and the deallocation function may or may not be called (it's unspecified), but the default deallocation functions are guaranteed to do nothing when passed a null pointer.

```suggestion
    delete [] index_covert_CSR2Triplet_;
    delete [] index_covert_extra_Diag2CSR_;
```

We do however have the pattern `if (ptr) delete ptr;` all over hiop, so it's up to @cnpetra if we should keep up the pattern.

cnpetra replied:
I am for using `delete ptr;` only as you pointe out. btw, I started using it quite some time ago.

ashermancinelli commented on src/LinAlg/hiopLinSolverSparseGinkgo.hpp:
```diff
@@ -0,0 +1,184 @@
+ * @ingroup LinearSolvers
+ */
+
+#ifndef FNAME
+#ifndef __bg__
+#define FNAME(f) f ## _
+#else
+#define FNAME(f) f
+#endif
+#endif
```
@cnpetra If memory serves, we have fortran name mangling macros defined elsewhere, don't we? If so, I think we should just reuse those to avoid potential conflict.

ashermancinelli commented on src/LinAlg/hiopLinSolverSparseGinkgo.hpp:
```diff
@@ -0,0 +1,184 @@
+
+namespace hiop {
+
+class hiopLinSolverIndefSparseGinkgo: public hiopLinSolverSymSparse
```
Will we even need matrix types other than `double` and `int`? If so, I think this should be a class template that just passes the template parameters to the `mtx` field.

pelesh replied:
HiOp has its own `index_type` and `size_type` for `int`s and these should be used. Using template parameters to specify scalar and index types for linear algebra objects would be a good long-term solution, but it would require a bit work.

ashermancinelli commented on src/Optimization/hiopKKTLinSysSparse.cpp:
```diff
@@ -660,6 +684,7 @@ namespace hiop
   hiopLinSolverNonSymSparse*
   hiopKKTLinSysSparseFull::determineAndCreateLinsys(const int &n, const int &n_con, const int &nnz)
   {
+      std::cout << "AM DOING THIS" << std::endl;
```
```suggestion
```

nychiang commented on src/LinAlg/hiopLinSolverSparseGinkgo.cpp:
```diff
@@ -0,0 +1,376 @@
+    int nnz_= mtx->get_num_stored_elements();
+    auto values = mtx->get_values();
+    int rowID_tmp{0};
+    for(int k=0;k<nnz_;k++){
+    values[k] = M_->M()[index_covert_CSR2Triplet_[k]];
+    }
```
Indentation

nychiang commented on src/LinAlg/hiopLinSolverSparseGinkgo.cpp:
```diff
@@ -0,0 +1,376 @@
+    for(int k=0;k<nnz_;k++){
+    values[k] = M_->M()[index_covert_CSR2Triplet_[k]];
+    }
+    for(int i=0;i<n_;i++){
+    if(index_covert_extra_Diag2CSR_[i] != -1)
+        values[index_covert_extra_Diag2CSR_[i]] += M_->M()[M_->numberOfNonzeros()-n_+i];
+    }
```
Indentation

nychiang commented on src/LinAlg/hiopLinSolverSparseGinkgo.hpp:
```diff
@@ -0,0 +1,184 @@
+
+  int nFakeNegEigs_;
+
+  std::shared_ptr<gko::Executor> exec;
+  std::shared_ptr<gko::matrix::Csr<double, int>> mtx;
+  std::shared_ptr<gko::LinOpFactory> reusable_factory;
+  std::shared_ptr<gko::LinOp> gko_solver;
```
Add "_" in the end of member variables.

nychiang commented on src/LinAlg/hiopLinSolverSparseGinkgo.cpp:
```diff
@@ -0,0 +1,376 @@
+      update_matrix(M_, mtx, index_covert_CSR2Triplet_, index_covert_extra_Diag2CSR_);
+    }
+
+    gko_solver = gko::share(reusable_factory->generate(mtx));
+    nlp_->runStats.linsolv.tmInertiaComp.start();
+    int negEigVal = nFakeNegEigs_;
+    nlp_->runStats.linsolv.tmInertiaComp.stop();
+    return negEigVal;
```
Can we add a list and some short explanations about the possible return flags from ginkgo?
For example, if ginkgo failed to factorize the matrix due to some numerical issue, we have something like `gko_solver.info[0] = -1`.

pelesh replied:
@cnpetra and @nychiang: Are there defined error codes for HiOp or for now we are using `-1` when a function fails?

nychiang commented on src/LinAlg/hiopLinSolverSparseGinkgo.hpp:
```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2017, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory (LLNL).
+// Written by Cosmin G. Petra, petra1@llnl.gov.
```
remove this line

nychiang reviewed (changes requested):
LGTM. It's better to rebase this branch to the latest `develop` branch and fix all the conflicts, before we merge it.

pelesh commented on src/LinAlg/hiopLinSolverSparseGinkgo.hpp:
```diff
@@ -0,0 +1,184 @@
+  void firstCall();
+//  virtual void diagonalChanged( int idiag, int extent );
+
+  void inline setFakeInertia(int nNegEigs)
+  {
+    nFakeNegEigs_ = nNegEigs;
+  }
```
I think you and @cnpetra should remove those functions from the code you implemented. The rest of us are just following your example :).

Kidding aside, this function does nothing and can be safely removed right away.

ashermancinelli commented:
Do we require CUDA when Ginkgo is enabled? Do we strictly require Ginkgo built with CUDA support when we build with Ginkgo? Trying to nail down the corner cases for our packaging code.

pelesh commented:
> Do we require CUDA when Ginkgo is enabled? Do we strictly require Ginkgo built with CUDA support when we build with Ginkgo? Trying to nail down the corner cases for our packaging code.

I would say no and no. Ginkgo has CPU, and multiple GPU backends (NVIDIA, AMD, Intel), so we should not require specific vendor support in order to enable Ginkgo with HiOp. The tests that @fritzgoebel added, however, should be enabled only if Ginkgo and CUDA are enabled.

fritzgoebel commented:
To try and answer the raised questions: I think the pull request is in a final(-ish) state.

BUT: To be clear, as mentioned in the PR description, the GLU branch and integration in Ginkgo only serves as a starting point for the direct solver design. Before going ahead merging this, let me make a few notes that arise from this experimental character:

- The GLU integration uses the openly available GLU code. This works on NVIDIA GPUs and contrary to the general design of Ginkgo, the choice of executor does not matter for the GLU solver. It will in any case attempt to use the GLU CUDA code.
- So, while you don't necessarily need to compile Ginkgo with the CUDA option enabled, you do need the CUDA toolkit available and it will be used whenever the Ginkgo solver is used.
- We are currently developing our own implementations of the several components of direct solvers. These will ultimately support all the different backends Ginkgo usually provides.

In conclusion, you can merge this pull request in order to move forward, but please do expect some changes on the Ginkgo side that will need to be merged at a later point.

pelesh reviewed (changes requested):
Before we merge this PR, we need to make sure Ginkgo tests run on all CI pipelines. Otherwise, the PR looks good.

ashermancinelli commented:
<details>

<summary>
This fails to build when I try to test it on Ascent:
</summary>

```console
[ 14%] Building CXX object src/LinAlg/CMakeFiles/hiopLinAlg.dir/hiopLinSolverSparseGinkgo.cpp.o
cd /gpfs/wolf/csc359/proj-shared/src/asher-hiop/spack-build-m3ljpl2/src/LinAlg && /gpfs/wolf/csc359/proj-shared/src/spack/lib/spack/env/gcc/g++ -DCAMP_HAVE_CUDA -I/gpfs/wolf/csc359/proj-shared/src/asher-hiop/spack-build-m3ljpl2 -I/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/Interface -I/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/Optimization -I/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg -I/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/Utils -I/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/_Internals -I/gpfs/wolf/csc359/proj-shared/src/spack/opt/spack/linux-rhel8-power9le/gcc-9.1.0/magma-2.6.2rc1-7gp5kh3vxd43n5fwwo4ignttsjsrepou/include -I/gpfs/wolf/csc359/proj-shared/src/spack/opt/spack/linux-rhel8-power9le/gcc-9.1.0/suite-sparse-5.10.1-ju7jhchvapbwgdkkrohatsyjfm23ybuf/include -I/gpfs/wolf/csc359/proj-shared/src/spack/opt/spack/linux-rhel8-power9le/gcc-9.1.0/metis-5.1.0-sg2ca2t5fi5e7hwu7e5cajwq6suvibx5/include -I/gpfs/wolf/csc359/proj-shared/src/spack/opt/spack/linux-rhel8-power9le/gcc-9.1.0/coinhsl-2015.06.23-qe3m7kwkfwxmm4wxabwatvurt6wvhhmt/include -isystem /autofs/nccsopen-svm1_sw/ascent/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.1.0/spectrum-mpi-10.4.0.3-20210112-6jbupg3thjwhsabgevk6xmwhd2bbyxdc/include -isystem /sw/ascent/spack-envs/base/opt/linux-rhel8-ppc64le/gcc-9.1.0/spectrum-mpi-10.4.0.3-20210112-6jbupg3thjwhsabgevk6xmwhd2bbyxdc/include -isystem /sw/ascent/cuda/11.4.2/include -isystem /gpfs/wolf/csc359/proj-shared/src/spack/opt/spack/linux-rhel8-power9le/gcc-9.1.0/camp-0.2.2-urpc3glcm5yiqlzngff5wpc65xitbgmz/include -isystem /gpfs/wolf/csc359/proj-shared/src/spack/opt/spack/linux-rhel8-power9le/gcc-9.1.0/umpire-6.0.0-3oxai6d4nfldsnf2udfddyghbsoartma/include -isystem /gpfs/wolf/csc359/proj-shared/src/spack/opt/spack/linux-rhel8-power9le/gcc-9.1.0/raja-0.14.0-sgbxrfcjwd4tnoi3syf57yuk47s2cvdh/include -isystem /gpfs/wolf/csc359/proj-shared/src/spack/opt/spack/linux-rhel8-power9le/gcc-9.1.0/cub-1.16.0-o5zdbepfpvsrwjqsgtdhopszb5m5k4rq/include -isystem /gpfs/wolf/csc359/proj-shared/src/spack/opt/spack/linux-rhel8-power9le/gcc-9.1.0/ginkgo-develop-2na5g46uycg75uda7virauhm2llytyo7/include -O2 -g -DNDEBUG -fPIC -pthread -fopenmp -std=gnu++14 -MD -MT src/LinAlg/CMakeFiles/hiopLinAlg.dir/hiopLinSolverSparseGinkgo.cpp.o -MF CMakeFiles/hiopLinAlg.dir/hiopLinSolverSparseGinkgo.cpp.o.d -o CMakeFiles/hiopLinAlg.dir/hiopLinSolverSparseGinkgo.cpp.o -c /gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp: In function 'std::shared_ptr<gko::matrix::Csr<double, int> > hiop::{anonymous}::transferTripletToCSR(std::shared_ptr<gko::Executor>, int, hiop::hiopMatrixSparse*, int**, int**)':
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:170:83: warning: narrowing conversion of 'n_' from 'int' to 'gko::dim<2>::dimension_type' {aka 'long unsigned int'} [-Wnarrowing]
  170 |     auto mtx = gko::share(gko::matrix::Csr<double, int>::create(exec, gko::dim<2>{n_, n_}, val_array, col_idxs, row_ptrs));
      |                                                                                   ^~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp: In function 'std::shared_ptr<gko::LinOpFactory> hiop::{anonymous}::setup_solver_factory(std::shared_ptr<const gko::Executor>, std::shared_ptr<gko::matrix::Csr<double, int> >)':
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:193:56: error: 'Mc64' is not a member of 'gko::reorder'
  193 |     auto preprocessing_fact = gko::share(gko::reorder::Mc64<double, int>::build().on(exec));
      |                                                        ^~~~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:193:61: error: expected primary-expression before 'double'
  193 |     auto preprocessing_fact = gko::share(gko::reorder::Mc64<double, int>::build().on(exec));
      |                                                             ^~~~~~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:193:69: error: expected primary-expression before 'int'
  193 |     auto preprocessing_fact = gko::share(gko::reorder::Mc64<double, int>::build().on(exec));
      |                                                                     ^~~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:196:51: error: 'Glu' is not a member of 'gko::factorization'; did you mean 'Ilu'?
  196 |     auto lu_fact = gko::share(gko::factorization::Glu<double, int>::build_reusable()
      |                                                   ^~~
      |                                                   Ilu
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:196:55: error: expected primary-expression before 'double'
  196 |     auto lu_fact = gko::share(gko::factorization::Glu<double, int>::build_reusable()
      |                                                       ^~~~~~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:196:63: error: expected primary-expression before 'int'
  196 |     auto lu_fact = gko::share(gko::factorization::Glu<double, int>::build_reusable()
      |                                                               ^~~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:214:53: error: 'ScaledReordered' is not a member of 'gko::solver'
  214 |     auto reusable_factory = gko::share(gko::solver::ScaledReordered<>::build()
      |                                                     ^~~~~~~~~~~~~~~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:214:69: error: expected primary-expression before '>' token
  214 |     auto reusable_factory = gko::share(gko::solver::ScaledReordered<>::build()
      |                                                                     ^
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:214:72: error: '::build' has not been declared
  214 |     auto reusable_factory = gko::share(gko::solver::ScaledReordered<>::build()
      |                                                                        ^~~~~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp: In member function 'virtual bool hiop::hiopLinSolverSymSparseGinkgo::solve(hiop::hiopVector&)':
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:285:74: warning: narrowing conversion of '((hiop::hiopLinSolverSymSparseGinkgo*)this)->hiop::hiopLinSolverSymSparseGinkgo::n_' from 'int' to 'gko::dim<2>::dimension_type' {aka 'long unsigned int'} [-Wnarrowing]
  285 |     auto dense_x = gko::matrix::Dense<double>::create(exec_, gko::dim<2>{n_, 1}, gko::Array<double>::view(exec_, n_, dx), 1);
      |                                                                          ^~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:286:74: warning: narrowing conversion of '((hiop::hiopLinSolverSymSparseGinkgo*)this)->hiop::hiopLinSolverSymSparseGinkgo::n_' from 'int' to 'gko::dim<2>::dimension_type' {aka 'long unsigned int'} [-Wnarrowing]
  286 |     auto dense_b = gko::matrix::Dense<double>::create(exec_, gko::dim<2>{n_, 1}, b_array, 1);
      |                                                                          ^~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp: In member function 'virtual bool hiop::hiopLinSolverNonSymSparseGinkgo::solve(hiop::hiopVector&)':
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:362:74: warning: narrowing conversion of '((hiop::hiopLinSolverNonSymSparseGinkgo*)this)->hiop::hiopLinSolverNonSymSparseGinkgo::n_' from 'int' to 'gko::dim<2>::dimension_type' {aka 'long unsigned int'} [-Wnarrowing]
  362 |     auto dense_x = gko::matrix::Dense<double>::create(exec_, gko::dim<2>{n_, 1}, x_array, 1);
      |                                                                          ^~
/gpfs/wolf/proj-shared/csc359/src/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:363:74: warning: narrowing conversion of '((hiop::hiopLinSolverNonSymSparseGinkgo*)this)->hiop::hiopLinSolverNonSymSparseGinkgo::n_' from 'int' to 'gko::dim<2>::dimension_type' {aka 'long unsigned int'} [-Wnarrowing]
  363 |     auto dense_b = gko::matrix::Dense<double>::create(exec_, gko::dim<2>{n_, 1}, b_array, 1);
      |                                                                          ^~
make[2]: *** [src/LinAlg/CMakeFiles/hiopLinAlg.dir/build.make:303: src/LinAlg/CMakeFiles/hiopLinAlg.dir/hiopLinSolverSparseGinkgo.cpp.o] Error 1
make[2]: Leaving directory '/gpfs/wolf/csc359/proj-shared/src/asher-hiop/spack-build-m3ljpl2'
make[1]: *** [CMakeFiles/Makefile2:355: src/LinAlg/CMakeFiles/hiopLinAlg.dir/all] Error 2
make[1]: Leaving directory '/gpfs/wolf/csc359/proj-shared/src/asher-hiop/spack-build-m3ljpl2'
make: *** [Makefile:149: all] Error 2
```

</details>

Has anyone in this thread built this branch with Ginkgo glu branch on a cuda platform? Have you been able to successfully build it?

cnpetra reviewed (commented):
thank you @fritzgoebel

Looks good to me minus a couple a couple minor issues

- [ ] codestyle issues
- [ ] needs rebase
- [ ] needs CI capable running GINKO (and the newly added GINKO examples)

I assume that compatibility with inertia free strategy of HiOp will be implemented in a follow up PR.

@pelesh I see that you are coordinating the effort re:  the three items above, please ping me when they are addressed and I will merge.

pelesh commented:
> I assume that compatibility with inertia free strategy of HiOp will be implemented in a follow up PR.

I believe inertia free strategy has been already implemented in this PR. 

The main showstopper is getting `ginkgo_solver` branch tests to run on CI pipeline and updating our spack packages to build HiOp with Ginkgo as a dependency.

ashermancinelli commented:
I'm almost ready to open a PR to enable CI for this branch. Where should that PR be since this PR is coming from Fritz's repo? @cnpetra

pelesh commented on src/Optimization/hiopKKTLinSysSparse.cpp:
```diff
@@ -295,15 +298,18 @@ namespace hiop
-                            neq+nineq);
+                            "KKT_SPARSE_XYcYd linsys: alloc GINKGO with matrix size %d (%d cons)\n",
+                            n, neq+nineq);
+          hiopLinSolverSymSparseGinkgo *p = new hiopLinSolverSymSparseGinkgo(n, nnz, nlp_);
+          p->setFakeInertia(neq + nineq);
+          linSys_ = p;        
```
Fake inertia is deprecated, can be safely removed from the code
```suggestion
          linSys_ = new hiopLinSolverSymSparseGinkgo(n, nnz, nlp_);
```

pelesh commented on src/Optimization/hiopKKTLinSysSparse.cpp:
```diff
@@ -612,13 +618,23 @@ namespace hiop
+        if( (nullptr == linSys_ && linear_solver == "auto") || linear_solver == "ginkgo") {
+          //ma57, pardiso and strumpack are not available or user requested ginkgo
+#ifdef HIOP_USE_GINKGO
+          hiopLinSolverSymSparseGinkgo *p = new hiopLinSolverSymSparseGinkgo(n, nnz, nlp_);
+          p->setFakeInertia(neq + nineq);
+          linSys_ = p;        
```
"Fake" inertia should not be used anymore:
```suggestion
          linSys_ = new hiopLinSolverSymSparseGinkgo(n, nnz, nlp_);
```

pelesh commented on src/LinAlg/hiopLinSolverSparseGinkgo.cpp:
```diff
@@ -0,0 +1,374 @@
+    nlp_->runStats.linsolv.tmInertiaComp.start();
+    int negEigVal = nFakeNegEigs_;
+    nlp_->runStats.linsolv.tmInertiaComp.stop();
+    return negEigVal;
```
Function `matrixChange` should return -1 if the matrix is singular and zero otherwise. The "fake inertia" approach is deprecated.

pelesh commented on src/LinAlg/hiopLinSolverSparseGinkgo.hpp:
```diff
@@ -0,0 +1,168 @@
+  int *index_covert_CSR2Triplet_;
+  int *index_covert_extra_Diag2CSR_;
+
+  int nFakeNegEigs_;
+
```
```suggestion
```
"Fake inertia" approach should not be used anymore and this member should be removed.

pelesh commented on src/LinAlg/hiopLinSolverSparseGinkgo.hpp:
```diff
@@ -0,0 +1,168 @@
+  virtual void firstCall();
+//  virtual void diagonalChanged( int idiag, int extent );
+
+  void inline setFakeInertia(int nFakeNegEigs)
+  {
+    nFakeNegEigs_ = nFakeNegEigs;
+  }
+
```
```suggestion
```
No need to keep `setFakeInertia` method anymore.

pelesh commented on src/LinAlg/hiopLinSolverSparseGinkgo.hpp:
```diff
@@ -0,0 +1,168 @@
+  int *index_covert_extra_Diag2CSR_;
+  std::unordered_map<int,int> extra_diag_nnz_map_;
+
+  int nFakeNegEigs_;
+ 
```
```suggestion
```
Just remove `nFakeNegEigs_` member.

ashermancinelli commented:
<details>

<summary>
Some build warnings I found:
</summary>

```console
/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp: In function ‘std::shared_ptr<gko::matrix::Csr<double, int> > hiop::{anonymous}::transferTripletToCSR(std::shared_ptr<gko::Executor>, int, hiop::hiopMatrixSparse*, int**, int**)’:
/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:170:89: warning: narrowing conversion of ‘n_’ from ‘int’ to ‘gko::dim<2>::dimension_type {aka long unsigned int}’ inside { } [-Wnarrowing]
     auto mtx = gko::share(gko::matrix::Csr<double, int>::create(exec, gko::dim<2>{n_, n_}, val_array, col_idxs, row_ptrs));
                                                                                         ^
/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp: In member function ‘virtual bool hiop::hiopLinSolverSymSparseGinkgo::solve(hiop::hiopVector&)’:
/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:285:74: warning: narrowing conversion of ‘((hiop::hiopLinSolverSymSparseGinkgo*)this)->hiop::hiopLinSolverSymSparseGinkgo::n_’ from ‘int’ to ‘gko::dim<2>::dimension_type {aka long unsigned int}’ inside { } [-Wnarrowing]
     auto dense_x = gko::matrix::Dense<double>::create(exec_, gko::dim<2>{n_, 1}, gko::Array<double>::view(exec_, n_, dx), 1);
                                                                          ^~
/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:286:74: warning: narrowing conversion of ‘((hiop::hiopLinSolverSymSparseGinkgo*)this)->hiop::hiopLinSolverSymSparseGinkgo::n_’ from ‘int’ to ‘gko::dim<2>::dimension_type {aka long unsigned int}’ inside { } [-Wnarrowing]
     auto dense_b = gko::matrix::Dense<double>::create(exec_, gko::dim<2>{n_, 1}, b_array, 1);
                                                                          ^~
/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp: In member function ‘virtual bool hiop::hiopLinSolverNonSymSparseGinkgo::solve(hiop::hiopVector&)’:
/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:362:74: warning: narrowing conversion of ‘((hiop::hiopLinSolverNonSymSparseGinkgo*)this)->hiop::hiopLinSolverNonSymSparseGinkgo::n_’ from ‘int’ to ‘gko::dim<2>::dimension_type {aka long unsigned int}’ inside { } [-Wnarrowing]
     auto dense_x = gko::matrix::Dense<double>::create(exec_, gko::dim<2>{n_, 1}, x_array, 1);
                                                                          ^~
/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverSparseGinkgo.cpp:363:74: warning: narrowing conversion of ‘((hiop::hiopLinSolverNonSymSparseGinkgo*)this)->hiop::hiopLinSolverNonSymSparseGinkgo::n_’ from ‘int’ to ‘gko::dim<2>::dimension_type {aka long unsigned int}’ inside { } [-Wnarrowing]
     auto dense_b = gko::matrix::Dense<double>::create(exec_, gko::dim<2>{n_, 1}, b_array, 1);
                                                                          ^~
/qfs/projects/exasgd/src/hiop/asher-hiop/src/Optimization/hiopKKTLinSysSparseCondensed.hpp(355): warning: variable "Jac_d" was declared but never referenced

/qfs/projects/exasgd/src/hiop/asher-hiop/src/Optimization/hiopKKTLinSysSparseCondensed.hpp(356): warning: variable "Hess" was declared but never referenced

/qfs/projects/exasgd/src/hiop/asher-hiop/src/Optimization/hiopKKTLinSysSparseCondensed.hpp(491): warning: variable "Dd_pert" was declared but never referenced

/qfs/projects/exasgd/src/hiop/asher-hiop/src/Optimization/hiopKKTLinSysSparseCondensed.hpp(492): warning: variable "Dx" was declared but never referenced

/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverCholCuSparse.hpp(75): warning: overloaded virtual function "hiop::hiopLinSolver::solve" is only partially overridden in class "hiop::hiopLinSolverCholCuSparse"

[ 30%] Building CUDA object src/LinAlg/CMakeFiles/hiopLinAlg.dir/hiopMatrixRajaDense.cpp.o
[ 31%] Building CUDA object src/LinAlg/CMakeFiles/hiopLinAlg.dir/hiopMatrixRajaSparseTriplet.cpp.o
/qfs/projects/exasgd/src/hiop/asher-hiop/src/LinAlg/hiopLinSolverCholCuSparse.hpp(75): warning: overloaded virtual function "hiop::hiopLinSolver::solve" is only partially overridden in class "hiop::hiopLinSolverCholCuSparse"
```
</details> 

<details>

<summary>
I also found the following tests to fail on Newell:
</summary>

```console
 ctest --output-on-failure -R NlpSparse7
Test project /qfs/projects/exasgd/src/hiop/asher-hiop/build
    Start 33: NlpSparse7_1
1/4 Test #33: NlpSparse7_1 .....................***Failed    3.39 sec
--------------------------------------------------------------------------
WARNING: There is at least non-excluded one OpenFabrics device found,
but there are no active ports detected (or Open MPI was unable to use
them).  This is most certainly not what you wanted.  Check your
cables, subnet manager configuration, etc.  The openib BTL will be
ignored for this job.

  Local host: newell03
--------------------------------------------------------------------------
LU solver from GINKGO library requires inertia free approach. Enabling now ...
===============
Hiop SOLVER
===============
Using 1 MPI ranks.
---------------
Problem Summary
---------------
Total number of variables: 500
     lower/upper/lower_and_upper bounds: 499 / 1 / 1
Total number of equality constraints: 2
Total number of inequality constraints: 499
     lower/upper/lower_and_upper bounds: 498 / 498 / 497
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
   0  6.4379656e+01 9.980e+00  1.010e+00   0.00  0.000e+00  0.000e+00  -(-)
Symbolic nonzero: 6498
Number of levels: 5
Device 0: Tesla V100-SXM2-16GB has been selected.
Gaussian elimination with static pivoting (GESP)...
1-Norm of A matrix is 501.002, Perturbation value is 5.01002e-06
Total GPU time: 1.51856 ms
[Warning] BiCGStab did NOT converged after 9 iters. The solution from iter 9 was returned.
         - Error code 1
         - Abs res=-nann         - Rel res=-nan
         - ||rhs||_2=22248.7   ||sol||_2=nan
[Warning] Requesting additional accuracy and stability from the KKT linear system at iteration 0 (safe mode ON) [2]
Device 0: Tesla V100-SXM2-16GB has been selected.
Gaussian elimination with static pivoting (GESP)...
1-Norm of A matrix is 501.002, Perturbation value is 5.01002e-06
Total GPU time: 1.61718 ms
[Warning] BiCGStab did NOT converged after 9 iters. The solution from iter 9 was returned.
         - Error code 1
         - Abs res=nann  - Rel res=nan
         - ||rhs||_2=nan   ||sol||_2=-nan
Minimum step size reached. The problem may be locally infeasible or the gradient inaccurate. Will try to restore feasibility.
nlpSparse_ex7.exe: /qfs/projects/exasgd/src/hiop/asher-hiop/src/Optimization/hiopIterate.cpp:381: virtual bool hiop::hiopIterate::takeStep_duals(const hiop::hiopIterate&, const hiop::hiopIterate&, const double&, const double&): Assertion `zl->matchesPattern(nlp->get_ixl())' failed.
[newell03:116483] *** Process received signal ***
[newell03:116483] Signal: Aborted (6)
[newell03:116483] Signal code:  (-6)
[newell03:116483] [ 0] [0x7fff847f04d8]
[newell03:116483] [ 1] /usr/lib64/libc.so.6(abort+0x2b4)[0x7fff76d21f94]
[newell03:116483] [ 2] /usr/lib64/libc.so.6(+0x355d4)[0x7fff76d155d4]
[newell03:116483] [ 3] /usr/lib64/libc.so.6(__assert_fail+0x64)[0x7fff76d156c4]
[newell03:116483] [ 4] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x103082d8]
[newell03:116483] [ 5] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x101a19bc]
[newell03:116483] [ 6] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x1019dd58]
[newell03:116483] [ 7] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x1019c430]
[newell03:116483] [ 8] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x1019a864]
[newell03:116483] [ 9] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x100a3e1c]
[newell03:116483] [10] /usr/lib64/libc.so.6(+0x25100)[0x7fff76d05100]
[newell03:116483] [11] /usr/lib64/libc.so.6(__libc_start_main+0xc4)[0x7fff76d052f4]
[newell03:116483] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 116483 on node newell03 exited on signal 6 (Aborted).
--------------------------------------------------------------------------

    Start 34: NlpSparse7_2
2/4 Test #34: NlpSparse7_2 .....................***Failed    1.95 sec
--------------------------------------------------------------------------
WARNING: There is at least non-excluded one OpenFabrics device found,
but there are no active ports detected (or Open MPI was unable to use
them).  This is most certainly not what you wanted.  Check your
cables, subnet manager configuration, etc.  The openib BTL will be
ignored for this job.

  Local host: newell03
--------------------------------------------------------------------------
===============
Hiop SOLVER
===============
Using 1 MPI ranks.
---------------
Problem Summary
---------------
Total number of variables: 500
     lower/upper/lower_and_upper bounds: 499 / 1 / 1
Total number of equality constraints: 2
Total number of inequality constraints: 499
     lower/upper/lower_and_upper bounds: 498 / 498 / 497
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
   0  6.4379656e+01 9.980e+00  1.010e+00   0.00  0.000e+00  0.000e+00  -(-)
Symbolic nonzero: 6498
Number of levels: 5
Device 0: Tesla V100-SXM2-16GB has been selected.
Gaussian elimination with static pivoting (GESP)...
1-Norm of A matrix is 501.002, Perturbation value is 5.01002e-06
Total GPU time: 1.53312 ms
[Warning] BiCGStab did NOT converged after 9 iters. The solution from iter 9 was returned.
         - Error code 1
         - Abs res=-nann         - Rel res=-nan
         - ||rhs||_2=22248.7   ||sol||_2=nan
[Warning] Requesting additional accuracy and stability from the KKT linear system at iteration 0 (safe mode ON) [2]
Device 0: Tesla V100-SXM2-16GB has been selected.
Gaussian elimination with static pivoting (GESP)...
1-Norm of A matrix is 501.002, Perturbation value is 5.01002e-06
Total GPU time: 1.40282 ms
[Warning] BiCGStab did NOT converged after 9 iters. The solution from iter 9 was returned.
         - Error code 1
         - Abs res=nann  - Rel res=nan
         - ||rhs||_2=nan   ||sol||_2=-nan
Minimum step size reached. The problem may be locally infeasible or the gradient inaccurate. Will try to restore feasibility.
nlpSparse_ex7.exe: /qfs/projects/exasgd/src/hiop/asher-hiop/src/Optimization/hiopIterate.cpp:381: virtual bool hiop::hiopIterate::takeStep_duals(const hiop::hiopIterate&, const hiop::hiopIterate&, const double&, const double&): Assertion `zl->matchesPattern(nlp->get_ixl())' failed.
[newell03:116515] *** Process received signal ***
[newell03:116515] Signal: Aborted (6)
[newell03:116515] Signal code:  (-6)
[newell03:116515] [ 0] [0x7fff8c2204d8]
[newell03:116515] [ 1] /usr/lib64/libc.so.6(abort+0x2b4)[0x7fff7e751f94]
[newell03:116515] [ 2] /usr/lib64/libc.so.6(+0x355d4)[0x7fff7e7455d4]
[newell03:116515] [ 3] /usr/lib64/libc.so.6(__assert_fail+0x64)[0x7fff7e7456c4]
[newell03:116515] [ 4] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x103082d8]
[newell03:116515] [ 5] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x101a19bc]
[newell03:116515] [ 6] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x1019dd58]
[newell03:116515] [ 7] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x1019c430]
[newell03:116515] [ 8] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x1019a864]
[newell03:116515] [ 9] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x100a3e1c]
[newell03:116515] [10] /usr/lib64/libc.so.6(+0x25100)[0x7fff7e735100]
[newell03:116515] [11] /usr/lib64/libc.so.6(__libc_start_main+0xc4)[0x7fff7e7352f4]
[newell03:116515] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 116515 on node newell03 exited on signal 6 (Aborted).
--------------------------------------------------------------------------

    Start 35: NlpSparse7_3
3/4 Test #35: NlpSparse7_3 .....................***Failed    0.44 sec
--------------------------------------------------------------------------
WARNING: There is at least non-excluded one OpenFabrics device found,
but there are no active ports detected (or Open MPI was unable to use
them).  This is most certainly not what you wanted.  Check your
cables, subnet manager configuration, etc.  The openib BTL will be
ignored for this job.

  Local host: newell03
--------------------------------------------------------------------------
===============
Hiop SOLVER
===============
Using 1 MPI ranks.
---------------
Problem Summary
---------------
Total number of variables: 500
     lower/upper/lower_and_upper bounds: 499 / 1 / 1
Total number of equality constraints: 2
Total number of inequality constraints: 499
     lower/upper/lower_and_upper bounds: 498 / 498 / 497
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
   0  6.4379656e+01 9.980e+00  1.010e+00   0.00  0.000e+00  0.000e+00  -(-)
nlpSparse_ex7.exe: /qfs/projects/exasgd/src/hiop/asher-hiop/src/Optimization/hiopKKTLinSysSparse.cpp:722: hiop::hiopLinSolverSymSparse* hiop::hiopKKTLinSysCompressedSparseXDYcYd::determineAndCreateLinsys(int, int, int, int): Assertion `linSys_&& "KKT_SPARSE_XDYcYd linsys: cannot instantiate backend linear solver"' failed.
[newell03:116540] *** Process received signal ***
[newell03:116540] Signal: Aborted (6)
[newell03:116540] Signal code:  (-6)
[newell03:116540] [ 0] [0x7fff7f6204d8]
[newell03:116540] [ 1] /usr/lib64/libc.so.6(abort+0x2b4)[0x7fff71b51f94]
[newell03:116540] [ 2] /usr/lib64/libc.so.6(+0x355d4)[0x7fff71b455d4]
[newell03:116540] [ 3] /usr/lib64/libc.so.6(__assert_fail+0x64)[0x7fff71b456c4]
[newell03:116540] [ 4] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x101e759c]
[newell03:116540] [ 5] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x101e5820]
[newell03:116540] [ 6] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x101a9c5c]
[newell03:116540] [ 7] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x101acacc]
[newell03:116540] [ 8] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x101999b4]
[newell03:116540] [ 9] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x100a3e1c]
[newell03:116540] [10] /usr/lib64/libc.so.6(+0x25100)[0x7fff71b35100]
[newell03:116540] [11] /usr/lib64/libc.so.6(__libc_start_main+0xc4)[0x7fff71b352f4]
[newell03:116540] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 116540 on node newell03 exited on signal 6 (Aborted).
--------------------------------------------------------------------------

    Start 36: NlpSparse7_4
4/4 Test #36: NlpSparse7_4 .....................***Failed    1.86 sec
--------------------------------------------------------------------------
WARNING: There is at least non-excluded one OpenFabrics device found,
but there are no active ports detected (or Open MPI was unable to use
them).  This is most certainly not what you wanted.  Check your
cables, subnet manager configuration, etc.  The openib BTL will be
ignored for this job.

  Local host: newell03
--------------------------------------------------------------------------
===============
Hiop SOLVER
===============
Using 1 MPI ranks.
---------------
Problem Summary
---------------
Total number of variables: 500
     lower/upper/lower_and_upper bounds: 499 / 1 / 1
Total number of equality constraints: 2
Total number of inequality constraints: 499
     lower/upper/lower_and_upper bounds: 498 / 498 / 497
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
   0  6.4379656e+01 9.980e+00  1.010e+00   0.00  0.000e+00  0.000e+00  -(-)
Symbolic nonzero: 6498
Number of levels: 5
Device 0: Tesla V100-SXM2-16GB has been selected.
Gaussian elimination with static pivoting (GESP)...
1-Norm of A matrix is 501.002, Perturbation value is 5.01002e-06
Total GPU time: 1.53594 ms
[Warning] BiCGStab did NOT converged after 9 iters. The solution from iter 9 was returned.
         - Error code 1
         - Abs res=-nann         - Rel res=-nan
         - ||rhs||_2=22248.7   ||sol||_2=nan
[Warning] Requesting additional accuracy and stability from the KKT linear system at iteration 0 (safe mode ON) [2]
Device 0: Tesla V100-SXM2-16GB has been selected.
Gaussian elimination with static pivoting (GESP)...
1-Norm of A matrix is 501.002, Perturbation value is 5.01002e-06
Total GPU time: 1.41786 ms
[Warning] BiCGStab did NOT converged after 9 iters. The solution from iter 9 was returned.
         - Error code 1
         - Abs res=nann  - Rel res=nan
         - ||rhs||_2=nan   ||sol||_2=-nan
Minimum step size reached. The problem may be locally infeasible or the gradient inaccurate. Will try to restore feasibility.
nlpSparse_ex7.exe: /qfs/projects/exasgd/src/hiop/asher-hiop/src/Optimization/hiopIterate.cpp:381: virtual bool hiop::hiopIterate::takeStep_duals(const hiop::hiopIterate&, const hiop::hiopIterate&, const double&, const double&): Assertion `zl->matchesPattern(nlp->get_ixl())' failed.
[newell03:116555] *** Process received signal ***
[newell03:116555] Signal: Aborted (6)
[newell03:116555] Signal code:  (-6)
[newell03:116555] [ 0] [0x7fff7e9104d8]
[newell03:116555] [ 1] /usr/lib64/libc.so.6(abort+0x2b4)[0x7fff70e41f94]
[newell03:116555] [ 2] /usr/lib64/libc.so.6(+0x355d4)[0x7fff70e355d4]
[newell03:116555] [ 3] /usr/lib64/libc.so.6(__assert_fail+0x64)[0x7fff70e356c4]
[newell03:116555] [ 4] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x103082d8]
[newell03:116555] [ 5] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x101a19bc]
[newell03:116555] [ 6] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x1019dd58]
[newell03:116555] [ 7] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x1019c430]
[newell03:116555] [ 8] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x1019a864]
[newell03:116555] [ 9] /qfs/projects/exasgd/src/hiop/asher-hiop/build/src/Drivers/nlpSparse_ex7.exe[0x100a3e1c]
[newell03:116555] [10] /usr/lib64/libc.so.6(+0x25100)[0x7fff70e25100]
[newell03:116555] [11] /usr/lib64/libc.so.6(__libc_start_main+0xc4)[0x7fff70e252f4]
[newell03:116555] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 116555 on node newell03 exited on signal 6 (Aborted).
--------------------------------------------------------------------------


0% tests passed, 4 tests failed out of 4

Total Test time (real) =   7.70 sec

The following tests FAILED:
         33 - NlpSparse7_1 (Failed)
         34 - NlpSparse7_2 (Failed)
         35 - NlpSparse7_3 (Failed)
         36 - NlpSparse7_4 (Failed)
Errors while running CTest
```
</details>

The branch is at least building now.

ashermancinelli commented:
To clarify, the branch was previously building because I had an issue with the dependencies which Nicholson and I figured out offline. It doesn't seem like the build failure was the fault of Ginkgo developers or an issue with this branch.

fritzgoebel commented:
Sorry for taking a while to respond - I was on vacation last week.
I rebased and addressed @pelesh's comments. Concerning the codestyle issues @cnpetra mentioned: would you mind pointing me to those?

cnpetra commented:
> Sorry for taking a while to respond - I was on vacation last week. I rebased and addressed @pelesh's comments. Concerning the codestyle issues @cnpetra mentioned: would you mind pointing me to those?

there are quite many in one of the source files... probably more efficient if I or @nychiang will quickly fix those than list each of the issues. Do you mind if we fix those? 

After that will merge the PR.

ashermancinelli commented:
@cnpetra @pelesh Changes required for this branch to be tested in continuous integration have not been merged into this branch yet.

pelesh commented:
> @cnpetra @pelesh Changes required for this branch to be tested in continuous integration have not been merged into this branch yet.

@ashermancinelli, can you make another quick PR?

ashermancinelli commented:
Yes, I'll try to get the PR ready as soon as possible.

ashermancinelli commented:
@cnpetra @pelesh  the following tests are failing on both newell and marianas when ginkgo is enabled:

```console
The following tests FAILED:
	 33 - NlpSparse7_1 (Failed)
	 34 - NlpSparse7_2 (Failed)
	 35 - NlpSparse7_3 (Failed)
	 36 - NlpSparse7_4 (Failed)
```

<details>

<summary>
Here's an example trace:
</summary>

```console
      Start 36: NlpSparse7_4
36: Test command: /share/apps/openmpi/4.1.0mlx5.0/gcc/9.1.0/bin/mpirun "-n" "1" "/people/svcexasgd/gitlab/94663/build/src/Drivers/nlpSparse_ex7.exe" "500" "-ginkgo" "-inertiafree" "-selfcheck"
36: Test timeout computed to be: 1800
36: [1651610240.184645] [dlt08:167600:0]       ib_iface.c:665  UCX  ERROR ibv_create_cq(cqe=4096) failed: Cannot allocate memory
36: [dlt08.local:167600] pml_ucx.c:273  Error: Failed to create UCP worker
36: ===============
36: Hiop SOLVER
36: ===============
36: Using 1 MPI ranks.
36: ---------------
36: Problem Summary
36: ---------------
36: Total number of variables: 500
36:      lower/upper/lower_and_upper bounds: 499 / 1 / 1
36: Total number of equality constraints: 2
36: Total number of inequality constraints: 499
36:      lower/upper/lower_and_upper bounds: 498 / 498 / 497
36: iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
36:    0  6.4379656e+01 9.980e+00  1.010e+00   0.00  0.000e+00  0.000e+00  -(-)
36: Symbolic nonzero: 6498
36: Number of levels: 5
36: Device 0: NVIDIA GeForce RTX 2080 Ti has been selected.
36: Gaussian elimination with static pivoting (GESP)...
36: 1-Norm of A matrix is 501.002, Perturbation value is 5.01002e-06
36: Total GPU time: 1.18682 ms
36: [Warning] BiCGStab did NOT converged after 9 iters. The solution from iter 9 was returned.
36: 	 - Error code 1
36: 	 - Abs res=nann	 - Rel res=nan
36: 	 - ||rhs||_2=22248.7   ||sol||_2=nan
36: [Warning] Requesting additional accuracy and stability from the KKT linear system at iteration 0 (safe mode ON) [2]
36: Device 0: NVIDIA GeForce RTX 2080 Ti has been selected.
36: Gaussian elimination with static pivoting (GESP)...
36: 1-Norm of A matrix is 501.002, Perturbation value is 5.01002e-06
36: Total GPU time: 0.973216 ms
36: [Warning] BiCGStab did NOT converged after 9 iters. The solution from iter 9 was returned.
36: 	 - Error code 1
36: 	 - Abs res=nann	 - Rel res=nan
36: 	 - ||rhs||_2=nan   ||sol||_2=nan
36: Minimum step size reached. The problem may be locally infeasible or the gradient inaccurate. Will try to restore feasibility.
36: nlpSparse_ex7.exe: /people/svcexasgd/gitlab/94663/src/Optimization/hiopIterate.cpp:381:
    virtual bool hiop::hiopIterate::takeStep_duals(const hiop::hiopIterate&, const hiop::hiopIterate&, const double&, const double&):
        Assertion `zl->matchesPattern(nlp->get_ixl())' failed.
36: [dlt08:167600] *** Process received signal ***
36: [dlt08:167600] Signal: Aborted (6)
36: [dlt08:167600] Signal code:  (-6)
36: [dlt08:167600] [ 0] /lib64/libpthread.so.0(+0xf630)[0x7fae613d4630]
36: [dlt08:167600] [ 1] /lib64/libc.so.6(gsignal+0x37)[0x7fae50ece387]
36: [dlt08:167600] [ 2] /lib64/libc.so.6(abort+0x148)[0x7fae50ecfa78]
36: [dlt08:167600] [ 3] /lib64/libc.so.6(+0x2f1a6)[0x7fae50ec71a6]
36: [dlt08:167600] [ 4] /lib64/libc.so.6(+0x2f252)[0x7fae50ec7252]
36: [dlt08:167600] [ 5] /people/svcexasgd/gitlab/94663/build/src/Drivers/nlpSparse_ex7.exe[0x6507c8]
36: [dlt08:167600] [ 6] /people/svcexasgd/gitlab/94663/build/src/Drivers/nlpSparse_ex7.exe[0x56e793]
36: [dlt08:167600] [ 7] /people/svcexasgd/gitlab/94663/build/src/Drivers/nlpSparse_ex7.exe[0x56bcd4]
36: [dlt08:167600] [ 8] /people/svcexasgd/gitlab/94663/build/src/Drivers/nlpSparse_ex7.exe[0x56a766]
36: [dlt08:167600] [ 9] /people/svcexasgd/gitlab/94663/build/src/Drivers/nlpSparse_ex7.exe[0x568f61]
36: [dlt08:167600] [10] /people/svcexasgd/gitlab/94663/build/src/Drivers/nlpSparse_ex7.exe[0x4aadab]
36: [dlt08:167600] [11] /lib64/libc.so.6(__libc_start_main+0xf5)[0x7fae50eba555]
36: [dlt08:167600] [12] /people/svcexasgd/gitlab/94663/build/src/Drivers/nlpSparse_ex7.exe[0x4a85f7]
36: [dlt08:167600] *** End of error message ***
36: --------------------------------------------------------------------------
36: Primary job  terminated normally, but 1 process returned
36: a non-zero exit code. Per user-direction, the job has been aborted.
36: --------------------------------------------------------------------------
36: --------------------------------------------------------------------------
36: mpirun noticed that process rank 0 with PID 167600 on node dlt08 exited on signal 6 (Aborted).
36: --------------------------------------------------------------------------
36/41 Test #36: NlpSparse7_4 ......................***Failed    7.41 sec
```
</details>

nychiang commented:
@ashermancinelli can you try the latest `develop` branch, which includes this branch already and some other fixes, to see if we have the same problem? Also let's try setting `ir_outer_maxit` to `0` to see if it converges.

pelesh commented:
> @ashermancinelli can you try the latest `develop` branch

Actually, all tests (including Ginkgo) pass in `develop` branch. HiOp recognizes it was built w/o Ginkgo, so it falls back to MA57 (what sort of defeats the purpose, but that's whole another issue).

@ashermancinelli, can you rebase branch with Ginkgo CI updates wrt `develop` and make a PR?

ashermancinelli commented:
I've pushed to branch `asher/ginkgo-ci-rebase-0-6-2`, we'll see what happens in ci.
